[Feature] Add a way to add user browser args while launching browser in android

pranavj1001 opened the pull request:
Closes https://github.com/microsoft/playwright/issues/19211

The code goes through the new param (`browserArgs: string[]`) and adds it to the existing list of arguments before launching the browser in android.

pranavj1001 committed:
add a new param browserArgs in BrowserContextOptions

pranavj1001 commented:
@microsoft-github-policy-service agree [company="BrowserStack"]

pranavj1001 commented:
@microsoft-github-policy-service agree company="BrowserStack"

mxschmitt commented on packages/playwright-core/src/server/android/android.ts:
```diff
@@ -264,12 +264,14 @@ export class AndroidDevice extends SdkObject {
       '--no-default-browser-check',
       `--remote-debugging-socket-name=${socketName}`,
       ...chromiumSwitches,
+      ...additionalBrowserArgs
```
```suggestion
      ...options?.args
```

mxschmitt reviewed (commented):
It should be just [here](https://github.com/microsoft/playwright/blob/e998b6cab94d1462192987537b924ef86153ea09/packages/protocol/src/protocol.yml#L3261) so it does not pollute the other structs.

pranavj1001 committed:
add code to use proxy and args in BrowserNewContextParams while launching browser in android device

pranavj1001 commented:
Hey @mxschmitt, I've done some changes in the code to check for `proxy` and `args` passed in options params of the `launchBrowser` method. Please have a look at it. 

I removed some code related to [headless](https://github.com/microsoft/playwright/blob/dbf3fc50896494d13153e45d08b75deae43afd7c/packages/playwright-core/src/server/chromium/chromium.ts#L294-L300), sandbox args and more. Also, what should be done with [playwrightOptions](https://github.com/microsoft/playwright/blob/dbf3fc50896494d13153e45d08b75deae43afd7c/packages/playwright-core/src/server/chromium/chromium.ts#L308). For now, I've removed it.

pranavj1001 committed:
fix lint issues

mxschmitt commented on docs/src/api/class-androiddevice.md:
```diff
@@ -143,6 +143,12 @@ Optional package name to launch instead of default Chrome for Android.
 * since: v1.9
 
+### option: AndroidDevice.launchBrowser.proxy = %%-browser-option-proxy-%%
+* since: v1.9
```
would be since 1.29

pranavj1001 replied:
Done

mxschmitt reviewed (commented):
Having a test for it would be nice, like we do for our normal proxy tests, one is enough for now.

pranavj1001 committed:
add test cases and minor changes in code

pranavj1001 commented on packages/playwright-core/src/server/android/android.ts:
```diff
@@ -264,17 +264,40 @@ export class AndroidDevice extends SdkObject {
+    return await this._connectToBrowser(socketName, options);
+  }
+
+  _defaultArgs(options: channels.BrowserNewContextParams, socketName: string): string[] {
```
Done

pranavj1001 committed:
remove console.log statements

pranavj1001 committed:
fix lint issue

pranavj1001 commented:
@mxschmitt, I've done the changes. Not sure why this [job](https://github.com/microsoft/playwright/actions/runs/3655257250/jobs/6176424542) failed. It had passed earlier when the code had a lint issue. Maybe rerunning that job will help.

mxschmitt commented on packages/protocol/src/protocol.yml:
```diff
@@ -867,6 +867,9 @@ Browser:
             bypass: string?
             username: string?
             password: string?
+        args:
```
lets add it instead to packages/protocol/src/protocol.yml line 3260 and use ´channels.AndroidDeviceLaunchBrowserParams´ as a type instead of NewContextOptions inside packages/playwright-core/src/server/android/android.ts

pranavj1001 replied:
Have done the changes. Please review it again.

pranavj1001 committed:
use AndroidDeviceLaunchBrowserParams instead of BrowserNewContextParams

mxschmitt commented:
Great patch, thanks!
